File: 977-squares-of-a-sorted-array/977-squares-of-a-sorted-array.js
/**
 * @param {number[]} nums
 * @return {number[]}
 */
// O(n) time | O (n) space
var sortedSquares = function(nums) {
    // declare squared array, initialized its value to 0
    const squaredArray = Array(nums.length).fill(0);
    // two pointer, left and right
    let left = 0;
    let right = nums.length - 1;
    for (let index = nums.length - 1; index >= 0; index--){
        const leftVal = nums[left];
        const rightVal = nums[right];
        if (Math.abs(leftVal) > Math.abs(rightVal)){
            // store square of leftVal, and increment left pointer
            squaredArray[index] = leftVal*leftVal;
            left++;
        }
        else{
            // store square of rightVal, and decrement right pointer
            squaredArray[index] = rightVal*rightVal;
            right--;
        }
    } 
    return squaredArray;
};

File: 977-squares-of-a-sorted-array/977-squares-of-a-sorted-array.py
class Solution:
    # O(n) time | O (n) space
    def sortedSquares(self, nums: List[int]) -> List[int]:
        # declare squared array, initialized its value to 0
        squaredArray = [0 for _ in nums]
        # two pointer, left and right
        left = 0
        right = len(nums) - 1
        for index in reversed(range(len(nums))):
            leftVal = nums[left]
            rightVal = nums[right]
            if abs(leftVal) > abs(rightVal):
                # store square of leftVal, and increment left pointer
                squaredArray[index] = leftVal*leftVal
                left += 1
            else:
                # store square of rightVal, and decrement right pointer
                squaredArray[index] = rightVal*rightVal
                right -= 1
        return squaredArray